database/rrdcalc.c

```diff
@@ -252,139 +315,176 @@ inline void rrdsetcalc_unlink(RRDCALC *rc) {
-    int ret = 0;
+static inline int rrdcalc_check_and_link_rrdset_callback(RRDSET *st, void *rrdcalc) {
+    RRDCALC *rc = rrdcalc;
 
-    // make sure it does not already exist
-    foreach_rrdcalc_in_rrdhost(host, rc) {
-        if (unlikely(rc->chart == chart_string && rc->name == name_string)) {
-            debug(D_HEALTH, "Health alarm '%s/%s' already exists in host '%s'.", chart, name, rrdhost_hostname(host));
-            info("Health alarm '%s/%s' already exists in host '%s'.", chart, name, rrdhost_hostname(host));
-            ret = 1;
-            break;
-        }
+    if(unlikely(rrdcalc_check_if_it_matches_rrdset(rc, st))) {
+        rrdcalc_link_to_rrdset(st, rc);
+        return -1;
     }
 
-    string_freez(name_string);
-    string_freez(chart_string);
-
-    return ret;
+    return 0;
 }
 
-inline uint32_t rrdcalc_get_unique_id(RRDHOST *host, STRING *chart, STRING *name, uint32_t *next_event_id) {
-    // re-use old IDs, by looking them up in the alarm log
-    ALARM_ENTRY *ae = NULL;
-    for(ae = host->health_log.alarms; ae ;ae = ae->next) {
-        if(unlikely(name == ae->name && chart == ae->chart)) {
-            if(next_event_id) *next_event_id = ae->alarm_event_id + 1;
-            break;
+// ----------------------------------------------------------------------------
+// RRDCALC rrdhost index management - constructor
+
+struct rrdcalc_constructor {
+    RRDHOST *rrdhost;                       // the host we operate upon
+    RRDCALC *from_config;                   // points to the original RRDCALC, as loaded from the config
+    RRDCALCTEMPLATE *from_rrdcalctemplate;  // the template this alert is generated from
+    RRDSET *rrdset;                         // when this comes from rrdcalctemplate, we have a matching rrdset
+    const char *overwrite_alert_name;       // when we have a dimension foreach, the alert is renamed
+    const char *overwrite_dimensions;       // when we have a dimension foreach, the dimensions filter is renamed
+
+    enum {
+        RRDCALC_REACT_NONE,
+        RRDCALC_REACT_NEW,
+    } react_action;
+};
+
+static void rrdcalc_rrdhost_insert_callback(const DICTIONARY_ITEM *item __maybe_unused, void *rrdcalc, void *constructor_data) {
+    RRDCALC *rc = rrdcalc;
+    struct rrdcalc_constructor *ctr = constructor_data;
+    RRDHOST *host = ctr->rrdhost;
+
+    rc->key = string_strdupz(dictionary_acquired_item_name(item));
+
+    if(ctr->from_rrdcalctemplate) {
+        rc->run_flags |= RRDCALC_FLAG_FROM_TEMPLATE;
+
+        RRDCALCTEMPLATE *rt = ctr->from_rrdcalctemplate;
+        RRDSET *st = ctr->rrdset;
+
+        rc->next_event_id = 1;
+        rc->name = (ctr->overwrite_alert_name) ? string_strdupz(ctr->overwrite_alert_name) : string_dup(rt->name);
+        rc->chart = string_dup(st->id);
+        uuid_copy(rc->config_hash_id, rt->config_hash_id);
+
+        rc->dimensions = (ctr->overwrite_dimensions) ? string_strdupz(ctr->overwrite_dimensions) : string_dup(rt->dimensions);
```

Comment:
We also have this suggestion:

```sh
*** CID 357074:  Concurrent data access violations  (MISSING_LOCK)
/database/rrdcalc.c: 435 in rrdcalc_rrdhost_insert_callback()
429             rc->dimensions = (ctr->overwrite_dimensions) ? string_strdupz(ctr->overwrite_dimensions) : string_dup(rt->dimensions);
430             rc->foreach_dimension = NULL;
431             rc->foreach_dimension_pattern = NULL;
432     
433             rc->green = rt->green;
434             rc->red = rt->red;
>>>     CID 357074:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "rc->value" without holding lock "netdata_rwlock_t.rwlock_t". Elsewhere, "rrdcalc.value" is accessed with "netdata_rwlock_t.rwlock_t" held 5 out of 6 times.
435             rc->value = NAN;
436             rc->old_value = NAN;
437     
438             rc->delay_up_duration = rt->delay_up_duration;
439             rc->delay_down_duration = rt->delay_down_duration;
440             rc->delay_max_duration = rt->delay_max_duration;

** CID 357073:    (RESOURCE_LEAK)
/libnetdata/dictionary/dictionary.c: 1123 in item_add_or_reset_value_and_acquire()
/libnetdata/dictionary/dictionary.c: 1160 in item_add_or_reset_value_and_acquire()
```

Reply:
The "Concurrent data access violations" found is a false positive. `rrdcalc_rrdhost_insert_callback()` is the constructor of the object. No other thread has access to it.

The `item_add_or_reset_value_and_acquire()` are accessing the value of a different object. Probably coverity got confused.